packages/lit-html/src/directive-helpers.ts

```diff
@@ -42,17 +48,23 @@ export const TemplateResultType = {
 export type TemplateResultType =
   (typeof TemplateResultType)[keyof typeof TemplateResultType];
 
+type isTemplateResultType = {
+  (val: unknown): val is TemplateResult | CompiledTemplateResult;
+  (val: unknown, type: TemplateResultType): val is TemplateResult;
```

Comment:
This just occurred to me as I was looking through, but wasn't part of the original implementation. Maybe we don't need this kind of granularity, so feel free to ignore.
```suggestion
  <T extends TemplateResultType>(val: unknown, type: T): val is TemplateResult<T>;
```

Reply:
Awesome!